papaparse.js

```diff
@@ -394,18 +394,29 @@
 			for (var row = 0; row < data.length; row++)
 			{
 				var maxCol = hasHeader ? fields.length : data[row].length;
-				var r = hasHeader ? fields : data[row];
 
-				if (skipEmptyLines !== 'greedy' || r.join('').trim() !== '')
+				var emptyLine = false;
+				var nullLine = hasHeader ? Object.keys(data[row]).length === 0 : data[row].length === 0;
+				if (skipEmptyLines && !hasHeader)
+					emptyLine = skipEmptyLines === 'greedy' ? data[row].join('').trim() === '' : data[row].length === 1 && data[row][0].length === 0;
+				if (skipEmptyLines === 'greedy' && hasHeader) {
+					var line = [];
+					for (var c = 0; c < maxCol; c++) {
+						var cx = dataKeyedByField ? fields[c] : c;
+						line.push(data[row][cx]);
+					}
+					emptyLine = line.join('').trim() === '';
+				}
+				if (!emptyLine)
 				{
 					for (var col = 0; col < maxCol; col++)
 					{
-						if (col > 0)
+						if (col > 0 && !nullLine)
 							csv += _delimiter;
 						var colIdx = hasHeader && dataKeyedByField ? fields[col] : col;
 						csv += safe(data[row][colIdx], col);
 					}
-					if (row < data.length - 1 && (!skipEmptyLines || maxCol > 0))
+					if (row < data.length - 1 && (!skipEmptyLines || maxCol > 0) && (!skipEmptyLines || !nullLine))
```

Comment:
You can rollup that last conditional group into the previous one:

```(!skipEmptyLines || maxCol > 0) && (!skipEmptyLines || !nullLine)```
`===`
```(!skipEmptyLines || maxCol > 0 || !nullLine)```

Reply:
Not quite equivalent. I wrote them rolled together like that originally and the tests fail (specifically when `skipEmptyLines` is `true` when unparsing with headers). We have two options as far as I can tell:

`(!skipEmptyLines || maxCol > 0) && (!skipEmptyLines || !nullLine)`

or

`(!skipEmptyLines || (maxCol > 0 && !nullLine))`

 If `skipEmptyLines === true` and `maxCol > 0 === true` then the newline will be added even if `nullLine === true` which is bad. We only want to check `maxCol > 0` and `!nullLine` if `skipEmptyLines === true` but both of them must be true in that case.

I thought the former was more readable, but we can do the latter if you prefer it.